File: frontend/src/api/marketUpdatesAdmin.ts
import { apiFetch } from "./client";

export type MarketAllowlistEntry = {
    id: number;
    phone_number: string;
    label: string | null;
    enabled: number;
    created_at: string;
    updated_at: string;
};

export type MarketInviteRequest = {
    id: number;
    phone_number: string;
    requested_label: string | null;
    message_text: string | null;
    status: "pending" | "approved" | "denied";
    created_at: string;
    updated_at: string;
};

export type MarketFeedbackEntry = {
    id: number;
    phone_number: string;
    feedback_text: string;
    source: string;
    created_at: string;
};

async function expectJson<T>(response: Response, fallbackError: string): Promise<T> {
    if (!response.ok) {
        const payload = await response.json().catch(() => ({}));
        throw new Error(payload.detail ?? fallbackError);
    }
    return response.json() as Promise<T>;
}

export async function fetchMarketAllowlist(): Promise<MarketAllowlistEntry[]> {
    const response = await apiFetch("/api/market-updates/admin/allowlist");
    return expectJson<MarketAllowlistEntry[]>(response, "Could not load allowlist");
}

export async function upsertMarketAllowlist(payload: {
    phone_number: string;
    label?: string;
    enabled?: boolean;
}): Promise<MarketAllowlistEntry> {
    const response = await apiFetch("/api/market-updates/admin/allowlist", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
    });
    return expectJson<MarketAllowlistEntry>(response, "Could not save allowlist entry");
}

export async function disableMarketAllowlist(phoneNumber: string): Promise<{ removed: boolean }> {
    const response = await apiFetch(`/api/market-updates/admin/allowlist/${encodeURIComponent(phoneNumber)}`, {
        method: "DELETE",
    });
    return expectJson<{ removed: boolean }>(response, "Could not disable allowlist entry");
}

export async function fetchMarketInviteRequests(status: string = "pending"): Promise<MarketInviteRequest[]> {
    const query = new URLSearchParams({ status }).toString();
    const response = await apiFetch(`/api/market-updates/admin/invite-requests?${query}`);
    return expectJson<MarketInviteRequest[]>(response, "Could not load invite requests");
}

export async function approveMarketInviteRequest(requestId: number): Promise<MarketInviteRequest> {
    const response = await apiFetch(`/api/market-updates/admin/invite-requests/${requestId}/approve`, {
        method: "POST",
    });
    return expectJson<MarketInviteRequest>(response, "Could not approve invite request");
}

export async function denyMarketInviteRequest(requestId: number): Promise<MarketInviteRequest> {
    const response = await apiFetch(`/api/market-updates/admin/invite-requests/${requestId}/deny`, {
        method: "POST",
    });
    return expectJson<MarketInviteRequest>(response, "Could not deny invite request");
}

export async function fetchMarketFeedback(limit: number = 200): Promise<MarketFeedbackEntry[]> {
    const query = new URLSearchParams({ limit: String(limit) }).toString();
    const response = await apiFetch(`/api/market-updates/admin/feedback?${query}`);
    return expectJson<MarketFeedbackEntry[]>(response, "Could not load feedback entries");
}

File: backend/app/routes/market_updates_admin.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.auth.dependencies import require_role
from market_updates.allowlist import (
    approve_invite_request,
    create_or_update_invite_request,
    deny_invite_request,
    disable_allowlist_number,
    list_allowlist,
    list_invite_requests,
    upsert_allowlist_number,
)
from market_updates.feedback_store import list_feedback_entries

router = APIRouter(prefix="/api/market-updates/admin", tags=["market-updates-admin"])


class AllowlistUpsertRequest(BaseModel):
    phone_number: str = Field(min_length=7, max_length=32)
    label: str | None = None
    enabled: bool = True


class InviteDraftRequest(BaseModel):
    phone_number: str = Field(min_length=7, max_length=32)
    requested_label: str | None = None
    message_text: str | None = None


@router.get("/allowlist")
def get_allowlist(_: dict = Depends(require_role("admin"))) -> list[dict]:
    return list_allowlist()


@router.post("/allowlist")
def post_allowlist(payload: AllowlistUpsertRequest, _: dict = Depends(require_role("admin"))) -> dict:
    try:
        return upsert_allowlist_number(payload.phone_number, label=payload.label, enabled=payload.enabled)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error


@router.delete("/allowlist/{phone_number}")
def delete_allowlist(phone_number: str, _: dict = Depends(require_role("admin"))) -> dict:
    removed = disable_allowlist_number(phone_number)
    return {"removed": removed}


@router.get("/invite-requests")
def get_invite_requests(status: str = "pending", _: dict = Depends(require_role("admin"))) -> list[dict]:
    return list_invite_requests(status=status)


@router.post("/invite-requests")
def post_invite_request(payload: InviteDraftRequest, _: dict = Depends(require_role("admin"))) -> dict:
    try:
        return create_or_update_invite_request(
            payload.phone_number,
            message_text=payload.message_text,
            requested_label=payload.requested_label,
        )
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error


@router.post("/invite-requests/{request_id}/approve")
def post_approve_request(request_id: int, _: dict = Depends(require_role("admin"))) -> dict:
    approved = approve_invite_request(request_id)
    if approved is None:
        raise HTTPException(status_code=404, detail="Invite request not found")
    return approved


@router.post("/invite-requests/{request_id}/deny")
def post_deny_request(request_id: int, _: dict = Depends(require_role("admin"))) -> dict:
    denied = deny_invite_request(request_id)
    if denied is None:
        raise HTTPException(status_code=404, detail="Invite request not found")
    return denied


@router.get("/feedback")
def get_feedback(limit: int = 200, _: dict = Depends(require_role("admin"))) -> list[dict]:
    return list_feedback_entries(limit=limit)
